designcompose/src/main/java/com/android/designcompose/ComputePaths.kt

```diff
@@ -69,26 +69,78 @@ class ComputedPaths(
     var strokeCap: androidx.compose.ui.graphics.StrokeCap?,
 )
 
-class ComputedShadow(val fills: List<Path>, val shadowStyle: BoxShadow)
+internal class ComputedShadow(val fills: List<Path>, val shadowStyle: BoxShadow)
+
+/// Maintain a cache of computed paths, as some paths and strokes are quite expensive to
+/// compute (if we need to provide a stroke for a path), causing `computePaths` to show
+/// up prominently in profiles for Squoosh without a cache.
+internal class ComputedPathCache {
+    internal class Entry(
+        val computedPaths: ComputedPaths,
+        // We could be more selective than using the entire style, and styles are
+        // immutable so we could also simply do object identity comparison rather
+        // than a deep equality comparison.
+        val viewShape: ViewShape,
+        val style: ViewStyle,
+        val density: Float,
+        val frameSize: Size,
+        val overrideSize: Size?,
+        val customArcAngle: Boolean,
+        val variableState: VariableState,
+    )
+
+    private var cache: HashMap<Int, Entry> = HashMap()
+    private var nextGeneration: HashMap<Int, Entry> = HashMap()
+
+    fun get(layoutId: Int): Entry? {
+        return cache[layoutId]
+    }
+
+    fun put(layoutId: Int, data: Entry) {
+        nextGeneration[layoutId] = data
+    }
+
+    fun collect() {
+        cache.clear()
+        cache.putAll(nextGeneration)
+        nextGeneration.clear()
+    }
+}
 
 internal fun ViewShape.computePaths(
     style: ViewStyle,
     density: Float,
     frameSize: Size,
     overrideSize: Size?,
     customArcAngle: Boolean,
-    vectorScaleX: Float,
-    vectorScaleY: Float,
     layoutId: Int,
     variableState: VariableState,
+    pathCache: ComputedPathCache
 ): ComputedPaths {
+    // Before we compute anything, see if this path didn't change since the last
+    // time.
+    val cacheEntry = pathCache.get(layoutId)
+    if (
+        cacheEntry != null &&
+            cacheEntry.viewShape == this &&
+            cacheEntry.style == style &&
+            cacheEntry.density == density &&
+            cacheEntry.frameSize == frameSize &&
+            cacheEntry.overrideSize == overrideSize &&
+            cacheEntry.customArcAngle == customArcAngle
+    ) {
+        // XXX: VariableState is never equal, even when the variables haven't changed.
```

Comment:
I remember having some issues with creation of VariableState, but I think it could be improved by putting it in a remember() with all the CompositionLocal objects that can cause variables to change.

Reply:
I've got an update coming for this bit -- I moved extraction of corner radius from variables up, and just include the computed border radii in the cache entry (then we get all of the factors for the generated path in the cache entry).